crates/holochain_persistence_lmdb/src/txn/lmdb.rs

```diff
@@ -0,0 +1,231 @@
+        eav_db: EavLmdbStorage<A>,
+        staging_cas_db: LmdbStorage,
+        staging_eav_db: EavLmdbStorage<A>,
+    ) -> Self {
+        Self {
+            cas_db,
+            eav_db,
+            staging_cas_db,
+            staging_eav_db,
+            phantom: std::marker::PhantomData,
+        }
+    }
+}
+
+fn to_api_error(e: rkv::error::StoreError) -> PersistenceError {
+    // Convert to lmdb persistence error
+    let e: crate::error::PersistenceError = e.into();
+
+    // Convert into api persistence error
+    let e: PersistenceError = e.into();
+    e
+}
+
+impl<A: Attribute> ContentAddressableStorage for EnvCursor<A> {
+    fn add(&self, content: &dyn AddressableContent) -> PersistenceResult<()> {
+        self.staging_cas_db.add(content)
+    }
+
+    fn contains(&self, address: &Address) -> PersistenceResult<bool> {
+        self.fetch(address)
+            .map(|maybe_content| maybe_content.is_some())
+    }
+
+    fn fetch(&self, address: &Address) -> PersistenceResult<Option<Content>> {
+        let maybe_content = self.staging_cas_db.fetch(address)?;
+
+        if maybe_content.is_some() {
+            return Ok(maybe_content);
+        }
+
+        let maybe_content = self.cas_db.fetch(address)?;
+
+        if let Some(content) = maybe_content {
+            self.staging_cas_db.add(&content)?;
+            Ok(Some(content))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn get_id(&self) -> uuid::Uuid {
+        self.cas_db.get_id()
+    }
+}
+
+impl<A: Attribute + serde::de::DeserializeOwned> EntityAttributeValueStorage<A> for EnvCursor<A> {
+    fn add_eavi(
+        &self,
+        eav: &EntityAttributeValueIndex<A>,
+    ) -> PersistenceResult<Option<EntityAttributeValueIndex<A>>> {
+        self.staging_eav_db.add_lmdb_eavi(eav).map_err(to_api_error)
+    }
+
+    fn fetch_eavi(
+        &self,
+        query: &EaviQuery<A>,
+    ) -> PersistenceResult<BTreeSet<EntityAttributeValueIndex<A>>> {
+        let eavis = self.staging_eav_db.fetch_eavi(query)?;
+
+        if !eavis.is_empty() {
+            return Ok(eavis);
+        }
+
+        let eavis = self.eav_db.fetch_eavi(query)?;
+
+        for eavi in &eavis {
+            self.staging_cas_db.add(eavi)?;
+        }
+        Ok(eavis)
+    }
+}
+
+impl<A: Attribute + DeserializeOwned> Cursor<A> for EnvCursor<A> {}
+
+#[derive(Clone)]
+pub struct LmdbCursorProvider<A: Attribute> {
+    cas_db: LmdbStorage,
+    eav_db: EavLmdbStorage<A>,
+    staging_path: PathBuf,
+    staging_initial_map_bytes: Option<usize>,
+}
+
+impl<A: Attribute + DeserializeOwned> CursorProvider<A> for LmdbCursorProvider<A> {
+    type Cursor = EnvCursor<A>;
+    fn create_cursor(&self) -> PersistenceResult<Self::Cursor> {
+        let staging_cas_db_name = format!("STAGING_CAS_{}", Uuid::new_v4());
+        let staging_eav_db_name = format!("STAGING_EAV_{}", Uuid::new_v4());
+        let db_names = vec![staging_cas_db_name.as_str(), staging_eav_db_name.as_str()];
```

Comment:
```suggestion
        let db_names = [&*staging_cas_db_name, &*staging_eav_db_name];
```
see below similar explanations